### DISCUSSION TITLE
io/buffer

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#152] [Announcements]
# io/buffer

Java程序员进阶之路，小白的零基础Java教程，Java缓冲流（Buffered）：读写速度有了质的飞升

https://tobebetterjavaer.com/io/buffer.html

**Comment by KayunZhang** [comment]
二哥，当int型作为Key时，hashMap偶尔会根据键来排序:
插入数据多了就不是有序的；
其他类型作为key又不是有序的。
我还在想这个hashMap不是无序的么，怎么也能用来排序，跟以前学的冲突了，怀疑了一下自我，查了一下才发现这个

**Comment by Snow-cold** [comment]
为什么03缓冲流联系那个 hashmap会排序，不是无序的嘛，不太理解

**Reply by yunsongqing** [reply]
1. 此处并未涉及hashmap的无序问题。

2. 为什么打印出来的是有顺序，是因为代码中使用了for循环来处理（主要因为打印的字符串内容本身每行最前面是有序号的），从1开始到map的size（     for (int i = 1; i <= lineMap.size(); i++)       ），这个地方来控制序号，同时根据这个序号去map中通过get取值，那么拼出来的字符串就是有序的了。

3. 这个地方并非传统的遍历map来拼接字符串。

**Comment by mking2023** [comment]
如果 getBufIfOpen()[pos++] 为 -118, 那么其原码表示为 `00000000 00000000 00000000 10001010`
请问这个原码是怎么计算出来的？

**Comment by AndiorSnow** [comment]
原码反码那里有问题
